16 Спринт

IvanPodlesnykh opened the pull request.

IvanPodlesnykh committed:
16 Спринт:
1. Сформировал структуру проекта (сделал пакеты для каждого экрана)
2. Реализовал шаблон MVVM для экрана проигрывателя

IvanPodlesnykh committed:
16 Спринт:
3. Заменил findViewById на ViewBinding на экранах поиска и проигрывателя
4. Реализовал шаблон MVVM для экрана поиска

IvanPodlesnykh committed:
16 Спринт:
5. Добавил в Creator прокидывание application в repository
6. Реализовал шаблон MVVM для экрана настроек

IvanPodlesnykh committed:
16 Спринт:
7. Реализовал шаблон MVVM для главного экрана (перенес навигацию в viewmodel)

ArturNurtdinov commented on app/src/main/java/com/ivanpodlesnykh/playlistmaker/data/search/SearchRepositoryImpl.kt:
```diff
@@ -0,0 +1,67 @@
+
+    private val musicService = retrofit.create(MusicApi::class.java)
+
+    private val searchHistory = SearchHistory(application.getSharedPreferences("sharedPref",
+        AppCompatActivity.MODE_PRIVATE
+    ))
```
SharedPreferences тоже можно в Creator вынести, в дальнейшем будет легче это отрефакторить

ArturNurtdinov commented on app/src/main/java/com/ivanpodlesnykh/playlistmaker/data/search/SearchRepositoryImpl.kt:
```diff
@@ -0,0 +1,67 @@
+    override fun searchTrack(request: String): Resource<List<Track>> {
+
+        if(!isConnected()) {
+            return Resource.Error("no_connection")
```
Это стоит вынести в константу, чтобы не копировать в двух местах, а лучше завести enum class и оперировать не строками, а целыми числами, т.к. их сравнение происходит быстрее, чем сравенение строк.

ArturNurtdinov commented on app/src/main/java/com/ivanpodlesnykh/playlistmaker/data/settings/SettingsRepositoryImpl.kt:
```diff
@@ -0,0 +1,39 @@
+    private val sharedPreferences = application.getSharedPreferences("sharedPref", Application.MODE_PRIVATE)
+    override fun getThemeSettings(): ThemeSettings {
+        val defaultTheme = (application.resources.configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK) == Configuration.UI_MODE_NIGHT_YES
+        if(sharedPreferences.getBoolean("DARK_THEME_KEY", defaultTheme)) {
```
Ключи для данных лучше вынести в отдельные константы

ArturNurtdinov commented on app/src/main/java/com/ivanpodlesnykh/playlistmaker/domain/search/api/SearchInteractor.kt:
```diff
@@ -0,0 +1,16 @@
+
+    fun clearSearchHistory()
+
+    interface SearchConsumer {
+        fun consume(foundTrackList: List<Track>, message: String?)
+    }
```
Уже есть интерфейс `Consumer<T>` в java sdk, можно воспользоваться им, чтобы не создавать лишний интерфейс

ArturNurtdinov commented on app/src/main/java/com/ivanpodlesnykh/playlistmaker/domain/sharing/impl/SharingInteractorImpl.kt:
```diff
@@ -0,0 +1,36 @@
+        return EmailData(arrayOf("ivanapodlesnykh@yandex.ru"),
+            "Сообщение разработчикам и разработчицам приложения Playlist Maker",
+            "Спасибо разработчикам и разработчицам за крутое приложение!")
+    }
```
Все строки должны быть вынесены в ресурсы, это одно из требований предыдущих спринтов

ArturNurtdinov commented on app/src/main/java/com/ivanpodlesnykh/playlistmaker/ui/player/activity/PlayerActivity.kt:
```diff
@@ -0,0 +1,143 @@
+        setContentView(binding.root)
+        requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
+
+        val track: Track = Gson().fromJson(intent.extras!!.getString("track"), Track::class.java)
```
Ключи для данных в константу лучше вынести, тем более это отвечает за данные, которые экран принимает на вход и если попытаться в другом месте открыть этот экран - то нужно будет читать весь код этого активити, чтобы понять, что нужно положить в аргументы данные с ключом `"track"`, а так, когда ключ в константе - это видно сразу

ArturNurtdinov commented on app/src/main/java/com/ivanpodlesnykh/playlistmaker/ui/player/activity/PlayerActivity.kt:
```diff
@@ -0,0 +1,143 @@
+
+    private fun handlePlayPauseButton(play: Boolean) {
+        val nightModeFlag = resources.configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK
+        if (nightModeFlag == Configuration.UI_MODE_NIGHT_YES) {
+            if(play) binding.playerPlayButton.setImageDrawable(getDrawable(R.drawable.player_play_button_night))
+            else binding.playerPlayButton.setImageDrawable(getDrawable(R.drawable.player_pause_button_night))
+        } else {
+            if(play) binding.playerPlayButton.setImageDrawable(getDrawable(R.drawable.player_play_button_daylight))
+            else binding.playerPlayButton.setImageDrawable(getDrawable(R.drawable.player_pause_button_daylight))
+        }
```
Можно не делать разные дроваблы для тёмной и светлой темы, для этого в самом xml файле иконки можно прописать ссылки на цвета, вместо захардкоженных там цветов, чтобы при переключении темы эти дроваблы перерисовались с правильными цветами

IvanPodlesnykh committed:
16 Спринт правки:
Добавил зависимость в SharingInteractorImpl от SettingsRepository, в SettingsRepositoryImpl реализовал функции, которые возвращают ссылку на пользовательское соглашение, сайт практикума и формируют EmailData для отправки письма разработчикам. Все значения подтягиваются из ресурсов.
